### ISSUE TITLE
Support for DNS-01 challenge with certbot

### ISSUE THREAD

**Issue by guorenxi** [caprover/caprover#1444]
Running caprover behind NAT with 80/443 blocked by ISP, and can not enable ssl due to the current challenge method(http-01), certbot have a lots of plugins for DNS-01 challenge which enable signing the certs by adding txt record to verify domain owner.

**Comment by maietta** [comment]
Hello @guorenxi,

This has been asked before, in different ways, including but not limited to supporting of wildcard SSL's. I'm actually developing a DNS plugin for Caprover but currently busy on paid projects. Until then, may I offer a workaround?

My scenario is that I am running Caprover on a Rapsberry Pi 4 device connected to my 4G LTE modem with unlimited internet for a 100% portable, battery operated "home lab" that goes with on travel. I am able to use it for email, web hosting with SSL's and wildcard certs. I'll be doing a full write-up on my company website as soon as I clear some of my paid projects, but in the teamtime, here's how it works:

I setup a cheap VPS at Vultr (https://www.vultr.com/?ref=7182700 is my referral link). The cheap VPS runs me $2.50/mo because I chose the cheapest plan I could find that includes IPv4 address. It's only purpose in life is to operate as a public IP entry point and reverse proxy to my Raspeberry Pi.

On the VPS I installed Haproxy and Wireguard as a server. I specify that I want 10.0.0.1 for my server IP on a new network interface wireguard creates.

On my portable Pi device, I installed Wireguard as a client. I specify that I want it's IP to be a static IP of 10.0.0.2.

On each device, I test that I can ping the other machine on those IP's. Once I confirm that it's working, I then configure wireguard to run at startup on both server and client.

Now for the reverse proxy.

On the cheap VPS, I install haproxy and add the following config for Caprover ports & Email services.

```
global
        log /dev/log    local0
        log /dev/log    local1 notice
        chroot /var/lib/haproxy
        stats socket /run/haproxy/admin.sock mode 660 level admin
        stats timeout 30s
        user haproxy
        group haproxy
        daemon

        # Default SSL material locations
        ca-base /etc/ssl/certs
        crt-base /etc/ssl/private

        # Default ciphers to use on SSL-enabled listening sockets.
        # For more information, see ciphers(1SSL).
        ssl-default-bind-ciphers kEECDH+aRSA+AES:kRSA+AES:+AES256:RC4-SHA:!kEDH:!LOW:!EXP:!MD5:!aNULL:!eNULL

defaults
        log     global
        mode    http
        option  httplog
        option  dontlognull
        timeout connect 5000
        timeout client  50000
        timeout server  50000
        errorfile 400 /etc/haproxy/errors/400.http
        errorfile 403 /etc/haproxy/errors/403.http
        errorfile 408 /etc/haproxy/errors/408.http
        errorfile 500 /etc/haproxy/errors/500.http
        errorfile 502 /etc/haproxy/errors/502.http
        errorfile 503 /etc/haproxy/errors/503.http
        errorfile 504 /etc/haproxy/errors/504.http

frontend http-in
    bind *:80
    mode tcp
    option tcplog
    option forwardfor
    timeout client  3h
    timeout server  3h
    option          clitcpka
    use_backend http-backend-80

backend http-backend-80
    mode tcp
    balance roundrobin
    server host0 10.0.0.2:443 check fall 3 rise 2

frontend https-in
    bind *:443
    log 127.0.0.1 local2
    mode tcp
    option tcplog
	option forwardfor
    timeout client  3h
    timeout server  3h
    option          clitcpka
    use_backend https-backend-443
    
backend https-backend-443
    mode tcp
    balance roundrobin
    server host0 10.0.0.2:443 check fall 3 rise 2
	
frontend ssh-in
    bind *:2222
    mode tcp
    option tcplog
    option forwardfor
    timeout client  3h
    timeout server  3h
    option          clitcpka
    use_backend ssh-backend

backend ssh-backend
    mode tcp
    balance roundrobin
    server host0 10.0.0.2:22 check fall 3 rise 2
	
frontend caprover-in
    bind *:3000
    mode tcp
    option tcplog
    option forwardfor
    timeout client  3h
    timeout server  3h
    option          clitcpka
    use_backend caprover-backend

backend caprover-backend
    mode tcp
    balance roundrobin
    server host0 10.0.0.2:3000 check fall 3 rise 2
	
frontend mail-25
    bind 104.207.155.3:25
    mode tcp
    option tcplog
    option forwardfor
    timeout client  3h
    timeout server  3h
    option          clitcpka
    use_backend mail-25-backend

backend mail-25-backend
    mode tcp
    balance roundrobin
    server host0 10.0.0.2:25 check fall 3 rise 2


frontend mail-587
    bind *:587
    mode tcp
    option tcplog
    option forwardfor
    timeout client  3h
    timeout server  3h
    option          clitcpka
    use_backend mail-587-backend

backend mail-587-backend
    mode tcp
    balance roundrobin
    server host0 10.0.0.2:587 check fall 3 rise 2
	
frontend mail-465
    bind *:465
    mode tcp
    option tcplog
    option forwardfor
    timeout client  3h
    timeout server  3h
    option          clitcpka
    use_backend mail-465-backend

backend mail-465-backend
    mode tcp
    balance roundrobin
    server host0 10.0.0.2:465 check fall 3 rise 2
	
frontend mail-143
    bind *:143
    mode tcp
    option tcplog
    option forwardfor
    timeout client  3h
    timeout server  3h
    option          clitcpka
    use_backend mail-143-backend

backend mail-143-backend
    mode tcp
    balance roundrobin
    server host0 10.0.0.2:143 check fall 3 rise 2
```

Once haproxy has been configured and is running, I can now moved to the final part, which is to setup Caprover.

When configuring Caprover during install (or later), my public IP becomes that of the network interface on the client side. That IP is 10.0.0.2 in my case.

With this setup, I am enjoying SSLs on my battery powered, portable home lab. My next step is to finish my plugin and share it with the world so people can use automatic DNS management with their favorite providers. I will be developing for Vultr DNS first. They offer a free Anycast DNS service that works amazingly well.

If you want more info or help, i'll be glad to share more info if you need it.

**Comment by guorenxi** [comment]
> 

Great idea, Thanks for the reply and the workaround, this is the best and might be the only option for those who don't have public ip address from their ISPs.

**Comment by guorenxi** [comment]
> Hello @guorenxi,
> 
> This has been asked before, in different ways, including but not limited to supporting of wildcard SSL's. I'm actually developing a DNS plugin for Caprover but currently busy on paid projects. Until then, may I offer a workaround?
> 
> My scenario is that I am running Caprover on a Rapsberry Pi 4 device connected to my 4G LTE modem with unlimited internet for a 100% portable, battery operated "home lab" that goes with on travel. I am able to use it for email, web hosting with SSL's and wildcard certs. I'll be doing a full write-up on my company website as soon as I clear some of my paid projects, but in the meantime, here's how it works:
> 
> I setup a cheap VPS at Vultr (https://www.vultr.com/?ref=7182700 is my referral link). The cheap VPS runs me $2.50/mo because I chose the cheapest plan I could find that includes IPv4 address. It's only purpose in life is to operate as a public IP entry point and reverse proxy to my Raspeberry Pi.
> 
> On the VPS I installed Haproxy and Wireguard as a server. I specify that I want 10.0.0.1 for my server IP on a new network interface wireguard creates.
> 
> On my portable Pi device, I installed Wireguard as a client. I specify that I want it's IP to be a static IP of 10.0.0.2.
> 
> On each device, I test that I can ping the other machine on those IP's. Once I confirm that it's working, I then configure wireguard to run at startup on both server and client.
> 
> Now for the reverse proxy.
> 
> On the cheap VPS, I install haproxy and add the following config for Caprover ports & Email services.
> 
> ```
> global
>         log /dev/log    local0
>         log /dev/log    local1 notice
>         chroot /var/lib/haproxy
>         stats socket /run/haproxy/admin.sock mode 660 level admin
>         stats timeout 30s
>         user haproxy
>         group haproxy
>         daemon
> 
>         # Default SSL material locations
>         ca-base /etc/ssl/certs
>         crt-base /etc/ssl/private
> 
>         # Default ciphers to use on SSL-enabled listening sockets.
>         # For more information, see ciphers(1SSL).
>         ssl-default-bind-ciphers kEECDH+aRSA+AES:kRSA+AES:+AES256:RC4-SHA:!kEDH:!LOW:!EXP:!MD5:!aNULL:!eNULL
> 
> defaults
>         log     global
>         mode    http
>         option  httplog
>         option  dontlognull
>         timeout connect 5000
>         timeout client  50000
>         timeout server  50000
>         errorfile 400 /etc/haproxy/errors/400.http
>         errorfile 403 /etc/haproxy/errors/403.http
>         errorfile 408 /etc/haproxy/errors/408.http
>         errorfile 500 /etc/haproxy/errors/500.http
>         errorfile 502 /etc/haproxy/errors/502.http
>         errorfile 503 /etc/haproxy/errors/503.http
>         errorfile 504 /etc/haproxy/errors/504.http
> 
> frontend http-in
>     bind *:80
>     mode tcp
>     option tcplog
>     option forwardfor
>     timeout client  3h
>     timeout server  3h
>     option          clitcpka
>     use_backend http-backend-80
> 
> backend http-backend-80
>     mode tcp
>     balance roundrobin
>     server host0 10.0.0.2:443 check fall 3 rise 2
> 
> frontend https-in
>     bind *:443
>     log 127.0.0.1 local2
>     mode tcp
>     option tcplog
> 	option forwardfor
>     timeout client  3h
>     timeout server  3h
>     option          clitcpka
>     use_backend https-backend-443
>     
> backend https-backend-443
>     mode tcp
>     balance roundrobin
>     server host0 10.0.0.2:443 check fall 3 rise 2
> 	
> frontend ssh-in
>     bind *:2222
>     mode tcp
>     option tcplog
>     option forwardfor
>     timeout client  3h
>     timeout server  3h
>     option          clitcpka
>     use_backend ssh-backend
> 
> backend ssh-backend
>     mode tcp
>     balance roundrobin
>     server host0 10.0.0.2:22 check fall 3 rise 2
> 	
> frontend caprover-in
>     bind *:3000
>     mode tcp
>     option tcplog
>     option forwardfor
>     timeout client  3h
>     timeout server  3h
>     option          clitcpka
>     use_backend caprover-backend
> 
> backend caprover-backend
>     mode tcp
>     balance roundrobin
>     server host0 10.0.0.2:3000 check fall 3 rise 2
> 	
> frontend mail-25
>     bind *:25
>     mode tcp
>     option tcplog
>     option forwardfor
>     timeout client  3h
>     timeout server  3h
>     option          clitcpka
>     use_backend mail-25-backend
> 
> backend mail-25-backend
>     mode tcp
>     balance roundrobin
>     server host0 10.0.0.2:25 check fall 3 rise 2
> 
> 
> frontend mail-587
>     bind *:587
>     mode tcp
>     option tcplog
>     option forwardfor
>     timeout client  3h
>     timeout server  3h
>     option          clitcpka
>     use_backend mail-587-backend
> 
> backend mail-587-backend
>     mode tcp
>     balance roundrobin
>     server host0 10.0.0.2:587 check fall 3 rise 2
> 	
> frontend mail-465
>     bind *:465
>     mode tcp
>     option tcplog
>     option forwardfor
>     timeout client  3h
>     timeout server  3h
>     option          clitcpka
>     use_backend mail-465-backend
> 
> backend mail-465-backend
>     mode tcp
>     balance roundrobin
>     server host0 10.0.0.2:465 check fall 3 rise 2
> 	
> frontend mail-143
>     bind *:143
>     mode tcp
>     option tcplog
>     option forwardfor
>     timeout client  3h
>     timeout server  3h
>     option          clitcpka
>     use_backend mail-143-backend
> 
> backend mail-143-backend
>     mode tcp
>     balance roundrobin
>     server host0 10.0.0.2:143 check fall 3 rise 2
> ```
> 
> Once haproxy has been configured and is running, I can now moved to the final part, which is to setup Caprover.
> 
> When configuring Caprover during install (or later), my public IP becomes that of the network interface on the client side. That IP is 10.0.0.2 in my case.
> 
> With this setup, I am enjoying SSLs on my battery powered, portable home lab. My next step is to finish my plugin and share it with the world so people can use automatic DNS management with their favorite providers. I will be developing for Vultr DNS first. They offer a free Anycast DNS service that works amazingly well.
> 
> If you want more info or help, i'll be glad to share more info if you need it.

Great idea, Thanks for the reply and the workaround, this is the best and might be the only option for those who don't have public ip address from their ISPs.

**Comment by maietta** [comment]
@guorenxi Not only that, but it doesn't require any open ports on router, etc. Traffic to and from your device is routed to your VPS through it's public IP.

**Comment by guorenxi** [comment]
> @guorenxi Not only that, but it doesn't require any open ports on router, etc. Traffic to and from your device is routed to your VPS through it's public IP.

Great. But DNS challenge could be more convenient since it doesn't require additional configurations. I looked into the code, looks like it is easy to implement, and I'm trying to find another workaround.

**Comment by maietta** [comment]
@guorenxi 

Hence why I'm developing a DNS plugin for Caprover. Since it's not a paid project, it's not a high priority item. If you come up with another solution, please let us know.

**Comment by alexander-potemkin** [comment]
I would also love to have an easy to use solution, for the cases when CapRover is behind the firewall, for the security purposes.

**Comment by maietta** [comment]
I got caught up in paid project for a bit longer but I promise I am still working on a plugin.

**Comment by asmaps** [comment]
Any news on this topic?

**Comment by jorgepsmatos** [comment]
Any news regarding this?

**Comment by githubsaturn** [comment]
This is one of the least upvoted feature requests. It won't be on a short term roadmap.

**Comment by adamshand** [comment]
I would also love this feature.  Being able to use Let's Encrypt certs on servers behind firewalls would be very useful.

As a side note, I think this is a duplicate of: https://github.com/caprover/caprover/issues/1761

**Comment by githubsaturn** [comment]
Thanks @adamshand ! 

I'll close this one so that people know which issue to upvote.

**Comment by maietta** [comment]
> @guorenxi
> 
> Hence why I'm developing a DNS plugin for Caprover. Since it's not a paid project, it's not a high priority item. If you come up with another solution, please let us know.

Oh yeah, I never finished working on that. I'll put that back in my calendar.
